torchtext/datasets/imdb.py

```diff
@@ -23,16 +28,25 @@
 @_add_docstring_header(num_lines=NUM_LINES, num_classes=2)
 @_create_dataset_directory(dataset_name=DATASET_NAME)
 @_wrap_split_argument(('train', 'test'))
-def IMDB(root, split):
-    def generate_imdb_data(key, extracted_files):
-        for fname in extracted_files:
-            *_, split, label, file = Path(fname).parts
-
-            if key == split and (label in ['pos', 'neg']):
-                with io.open(fname, encoding="utf8") as f:
-                    yield label, f.read()
-    dataset_tar = download_from_url(URL, root=root,
-                                    hash_value=MD5, hash_type='md5')
-    extracted_files = extract_archive(dataset_tar)
-    iterator = generate_imdb_data(split, extracted_files)
-    return _RawTextIterableDataset(DATASET_NAME, NUM_LINES[split], iterator)
+def IMDB(root: str, split: Union[Tuple[str], str]):
+    if not is_module_available("torchdata"):
+        raise ModuleNotFoundError("Package `torchdata` not found. Please install following instructions at `https://github.com/pytorch/data`")
+
+    url_dp = IterableWrapper([URL])
+
+    cache_dp = url_dp.on_disk_cache(
+        filepath_fn=lambda x: os.path.join(root, _PATH),
+        hash_dict={os.path.join(root, _PATH): MD5}, hash_type="md5"
+    )
+    cache_dp = HttpReader(cache_dp).end_caching(mode="wb", same_filepath_fn=True)
+    cache_dp = FileOpener(cache_dp, mode="b")
+
+    extracted_files = cache_dp.read_from_tar()
+
+    def filter_imdb_data(key, fname):
+        *_, split, label, file = Path(fname).parts
+        return key == split and (label in ['pos', 'neg'])
```

Comment:
In the current implementation, we construct the `['pos', 'neg']` list for every file contained within `extracted_files` datapipe. 

Instead, can we create a `labels` set outside the `filter_imdb_data()` function and then do a contains check to see if the label exists within the `labels` set? This would make it so the set is only created once regardless of the number of files within `extracted_files`. Also, `contains` operations are faster on a set than on a list in Python.

Reply:
Sets are asymptotically faster with `__contains__`, but with a len-2 iterable, they're about the same. Interestingly, I suspected checking `(label == "pos") or (label == "neg")` would be faster, but it's slower on average...

```python
In [1]: def f1(label):
   ...:     return label in ["pos", "neg"]

In [2]: def f2(label):
   ...:     return label in {"pos", "neg"}

In [3]: def f3(label):
   ...:     return (label == "pos") or (label == "neg")

In [4]: %timeit f1("pos")
90.9 ns ± 1.99 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [5]: %timeit f1("neg")
109 ns ± 1.85 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [6]: %timeit f2("pos")
96 ns ± 2.38 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [7]: %timeit f2("neg")
94.2 ns ± 2.55 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [8]: %timeit f3("pos")
100 ns ± 3.85 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [9]: %timeit f3("neg")
120 ns ± 2.84 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
```

Reply:
Thanks @erip for helping verify this! You're right that there is a negligible difference in performance when the length of the iterable is small. I just thought it would be better practice but I guess this would be more of a nit comment 😃